src/audio/mixin_mixout/mixin_mixout_hifi3.c

```diff
@@ -0,0 +1,533 @@
+				  uint8_t source_channel_index, uint8_t source_channel_count,
+				  uint32_t frame_count, uint16_t gain)
+{
+	ae_int16 *dst, *src;
+	int frames_to_mix, frames_to_copy, left_frames;
+	int n, nmax, frames, i, shift;
+	int inoff = source_channel_count * sizeof(ae_int16);
+	int outoff = sink_channel_count * sizeof(ae_int16);
+	ae_int16x4 in;
+	ae_int16x4 out;
+	ae_int16 *pgain = (ae_int16 *)&gain;
+	ae_int16x4 gain_v;
+	ae_int32x2 temp, out1;
+
+	dst = (ae_int16 *)sink->w_ptr + start_frame * sink_channel_count + sink_channel_index;
+	src = (ae_int16 *)source->r_ptr + source_channel_index;
+	src = audio_stream_wrap(source, src);
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+	shift = IPC4_MIXIN_GAIN_SHIFT + 1;
+
+	/* store gain to a AE_DR register gain_v*/
+	AE_L16_IP(gain_v, pgain, 0);
+
+	/* set source as circular buffer, hifi3 only have 1 circular buffer*/
+	AE_SETCBEGIN0(source->addr);
+	AE_SETCEND0(source->end_addr);
+
+	for (left_frames = frames_to_mix; left_frames; left_frames -= frames) {
+		/* audio_stream_wrap() is required and is done below in a loop */
+		dst = audio_stream_wrap(sink, dst);
+
+		/* calculate the remaining samples of sink*/
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(left_frames * sink_channel_count, nmax);
+
+		/* frames is the processed frame count in this loop*/
+		frames = 0;
+		for (i = 0; i < n; i += source_channel_count) {
+			AE_L16_XC(in, src, inoff);
+			AE_L16_XP(out, dst, 0);
+			/* Q1.15 * Q1.15 to Q2.30*/
+			temp = AE_MULF16SS_00(in, gain_v);
+			temp = AE_SRAA32(temp, shift);
+			out1 = AE_SEXT32X2D16_10(out);
+			temp = AE_ADD32S(temp, out1);
+			temp = AE_SLAA32S(temp, 16);
+			out = AE_ROUND16X4F32SSYM(temp, temp);
+			AE_S16_0_XP(out, dst, outoff);
+			frames++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames; left_frames -= frames) {
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(left_frames * sink_channel_count, nmax);
+		frames = 0;
+
+		for (i = 0; i < n; i += source_channel_count) {
+			AE_L16_XC(in, src, inoff);
+			AE_L16_XP(out, dst, 0);
+			temp = AE_MULF16SS_00(in, gain_v);
+			temp = AE_SRAA32(temp, shift);
+			temp = AE_SLAA32S(temp, 16);
+			out = AE_ROUND16X4F32SSYM(temp, temp);
+			AE_S16_0_XP(out, dst, outoff);
+			frames++;
+		}
+	}
+}
+
+static void mute_channel_s16(struct audio_stream __sparse_cache *stream, uint8_t channel_index,
+			     uint32_t start_frame, uint32_t mixed_frames, uint32_t frame_count)
+{
+	uint32_t skip_mixed_frames, left_frames;
+	int off = stream->channels * sizeof(ae_int16);
+	ae_int16 *ptr;
+	ae_int16x4 zero = AE_ZERO16();
+
+	assert(mixed_frames >= start_frame);
+	skip_mixed_frames = mixed_frames - start_frame;
+
+	if (frame_count <= skip_mixed_frames)
+		return;
+	frame_count -= skip_mixed_frames;
```

Comment:
No point to use unsigned types when doing subtracting them etc. The DSP ALU computes signed integers natively so why not use them.